Draw 3D gridlines below axis lines, labels, text, and ticks

scottshambaugh opened the pull request:
## PR Summary
Right now, 3D gridlines are drawn on top of more important elements such as the axis lines, labels, text, and ticks. I consider this a visual bug, as it obscures the information that the user cares about. This PR changes the draw order so that 3D gridlines are above only the 3D axis panes, and below everything else.

This requires updating 53 out of 57 test_axes3d baseline images, all 3 test_label3d baseline images, and 2 other baseline images that used 3D axes. These are minor changes in all instances, and I think these two diffs are the largest & most illustrative (you'll probably want to look at the "Files changed" swipe view for these):

axes3d_primary_views:
![axes3d_primary_views-failed-diff](https://user-images.githubusercontent.com/14363975/225811384-b7c7afcf-1f06-48d4-bdd5-619200cc4469.png)

axes3d_labelpad:
![axes3d_labelpad-failed-diff](https://user-images.githubusercontent.com/14363975/225812558-9482ea24-a3a5-4ff5-bacf-549db50cd321.png)

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [N/A] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

QuLogic commented:
Have you confrmed that this looks correct when grid lines are disabled? We don't appear to have any tests that turn them off, so I'm uncertain if that's even possible for 3D Axes.

oscargus commented:
Oh, btw, can you please change all tests that now require new images to use `style='mpl20'` (if they are not)?

scottshambaugh commented:
Thanks for the review! @QuLogic I added a new test to exercise hiding gridlines, ticks, and axis spines, and @oscargus I added `style='mpl20'` to all the applicable image tests. There were a few that had `style=('classic', '_classic_test_patch')`, which I left alone.

oscargus commented:
> There were a few that had `style=('classic', '_classic_test_patch')`, which I left alone.

That is just there to enable a transition to `mpl20` as default at a later stage, see https://github.com/matplotlib/matplotlib/issues/24716, so please update those as well.

scottshambaugh commented:
Would recommend merging this sooner rather than later, since it will conflict with any MR that modifies the mplot3d test images.

QuLogic commented on lib/mpl_toolkits/mplot3d/tests/test_axes3d.py:
```diff
@@ -39,7 +39,19 @@ def test_invisible_axes(fig_test, fig_ref):
+def test_invisible_grid_ticks_axis():
+    fig = plt.figure()
+    ax = fig.add_subplot(projection='3d')
+    ax.grid(False)
+    ax.set_xticks([])
+    ax.set_yticks([])
+    ax.set_zticks([])
```
If I'm not mistaken, this might be a bit too much, as if there are no tick locations, then there's nowhere to draw the grid, so it doesn't quite check that turning off _the grid itself_ worked.

And if that's true, we may not need a full image comparison here, but instead could do a `check_figures_equal` test between setting no ticks vs turning off the grid. Though you might have to fiddle with the tick markers themselves then, so it may not be viable.

scottshambaugh replied:
That's a great catch, I just tried that out and you're right! Split the new test into two, should actually be testing what we want it to now.
